lib/eco/treelexer/lexer.py

```diff
@@ -80,24 +88,54 @@ def match_one(self, pattern):
         self.exactmatch = False
         return False
 
-    def match_star(self, pattern):
-        while self._inner_match(pattern.c):
-             pass
-        return True
+    def match_star(self, pattern, l=None):
+        if l is None or len(l) == 1:
+            while self._inner_match(pattern.c):
+                pass
+            return True
+
+        if pattern.ng: # non-greedy
+            while True:
+                tmp = self.save_state()
+                if self.match_list(l[1:]):
+                    return True
+                self.load_state(tmp)
+                if self._inner_match(pattern.c):
+                    continue
+                return False
+        else: # greedy
+            tmp = self.save_state()
+            if self._inner_match(pattern.c) and self.match_list(l):
+                return True
+            # remember how far we looked ahead before backtracking
+            etmp = (self.la, self.exactmatch)
+            self.load_state(tmp)
+            self.la, self.exactmatch = etmp
+            if self.match_list(l[1:]):
```

Comment:
Ok, so this bit is confusing. I expected the greedy backtracing to count down negatively. So let's say we greedily matched 5 characters, we'd then backtrack and try 4, then 3 (etc.). This looks like it's counting up from 0?

Reply:
This only removes the star regex from the list and matches the rest. E.g. if the regex is `a*aa`, it first matches greedily, which then fails above so we backtrack, remove `a*` and match the remaining `aa`.

Reply:
If you want to give me a test I can check if it fails.

Reply:
Note that line 108 recursively matches star again as it is contained within `l`. So it says something like `if match(a*) and match(a*aaa)`

Reply:
Can you be a bit clearer what the `l[1:]` bit is doing? I'm not sure I've understood either of the explanations so far correctly.

Reply:
It removes the star regex from the list of regex's it still has to match. Let's `match_star` starts with something like `a*aa`. Then `pattern.c` is `a*`, while `l` is the remaining bit `aa`. We then try to recursively do `match_star(a*) and match(a*aa)`, which greedily matches as many `a`'s as it can until it fails because there are no more `a`'s to match. This means it has matched too many `a`'s using `a*` so there is nothing left to match the remaining bit `aa`. So it backtracks until it can match `l[1:]` which is just `aa`.

I agree this is a bit confusing due to the somewhat hidden recursion, but I couldn't find an easier way of doing this.

Reply:
OK, so I get the `pattern.c` thing now; but I don't see where the backtracking happens?

Reply:
The backtracking happens when the remainder of the pattern `l` can't be matched in line 114. This then returns False, which makes `match(l)` 108 fail, which then backtracks pattern.c once (i.e. one `a`) and then tries to match `aa` again, if that still fails, it returns false, which backtracks another `a` and so on, until line 114 succeeds.

Reply:
Maybe I'm dumb, but backtracing should involve a `- 1` somewhere, I would assume? And I just don't see that. I feel like I'm missing out on something obvious.

Reply:
Oh I see what you mean now! This happens within `self.load_state(tmp)`, which resets the `self.pos` (the position how far we've matched) to before a single `a` was matched. Due to the recursion this then happens multiple times until we are back at a `pos` where `match(l[1:])` can succeed.

Reply:
OK, so I feel a bit dumb: `load_state` doesn't fully reset the lexing state back to what it was at `save_state`?

Reply:
Yes, that sounds correct.

Reply:
OK: we can't do this, I'm afraid, because users will write perfectly normal content (e.g. a long string) that will blow up the stack. [I originally thought your recursion depth was O(m) where *m* is the number of elements in the regex. That's a standard-ish way of implementing regexs.]

Reply:
I agree, I tried to make this non-recursive but gave up after a while. I'll have another jab at this, but if it takes too long I'll fix it it another PR.

Reply:
Okay, have a look at the latest commit, which is a non-recursive solution for this.